File: libs/moqtail-rs/src/model/common/location.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use bytes::{Bytes, BytesMut};

use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Location {
  pub group: u64,
  pub object: u64,
}

impl Location {
  pub fn new(group: u64, object: u64) -> Self {
    Self { group, object }
  }

  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(self.group)?;
    buf.put_vi(self.object)?;
    Ok(buf.freeze())
  }

  pub fn deserialize(bytes: &mut Bytes) -> Result<Self, ParseError> {
    let group = bytes.get_vi()?;
    let object = bytes.get_vi()?;
    Ok(Location { group, object })
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_roundtrip() {
    let original = Location::new(1, 100);
    let mut bytes = original.serialize().unwrap();
    let deserialized = Location::deserialize(&mut bytes).unwrap();
    assert_eq!(original, deserialized);
  }

  #[test]
  fn test_ordering() {
    let loc1 = Location::new(1, 1);
    let loc2 = Location::new(1, 2);
    let loc3 = Location::new(2, 1);
    assert!(loc1 < loc3);
    assert!(loc2 < loc3);
    assert!(loc1 < loc2);
  }

  #[test]
  fn test_deserialize_error() {
    let mut bytes = BytesMut::new();
    bytes.put_vi(1).unwrap(); // Only group, missing object
    let result = Location::deserialize(&mut bytes.freeze());
    assert!(result.is_err());
  }
}

File: libs/moqtail-ts/src/model/common/location.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from './byte_buffer'
import { NotEnoughBytesError } from '../error/error'

/**
 * @public
 * Represents a position in a MOQT track, consisting of a group and object index.
 *
 * Used for specifying start/end positions in subscription and fetch requests.
 * - `group`: The group index (e.g. segment, GOP, or logical group).
 * - `object`: The object index within the group (e.g. frame, chunk).
 */
export class Location {
  /**
   * The group index for this location.
   */
  public readonly group: bigint

  /**
   * The object index within the group for this location.
   */
  public readonly object: bigint

  /**
   * Constructs a new Location.
   * @param group - The group index (number or bigint).
   * @param object - The object index (number or bigint).
   */
  constructor(group: bigint | number, object: bigint | number) {
    this.group = BigInt(group)
    this.object = BigInt(object)
  }

  /**
   * Serializes this Location to a FrozenByteBuffer.
   * @returns The serialized buffer.
   * @throws CastingError if group or object is negative.
   * @throws VarIntOverflowError if group or object exceeds varint encoding limits.
   */
  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(this.group)
    buf.putVI(this.object)
    return buf.freeze()
  }

  /**
   * Deserializes a Location from a buffer.
   * @param buf - The buffer to read from.
   * @returns The deserialized Location.
   * @throws NotEnoughBytesError if buffer does not contain enough bytes.
   */
  static deserialize(buf: BaseByteBuffer): Location {
    const group = buf.getVI()
    const object = buf.getVI()
    return new Location(group, object)
  }

  /**
   * Checks if this Location is equal to another.
   * @param other - The other Location to compare.
   * @returns True if both group and object are equal.
   */
  equals(other: Location): boolean {
    return this.group === other.group && this.object === other.object
  }

  /**
   * Compares this Location to another for ordering.
   * @param other - The other Location to compare.
   * @returns -1 if this \< other, 1 if this \> other, 0 if equal.
   */
  compare(other: Location): number {
    if (this.group < other.group) return -1
    if (this.group > other.group) return 1
    if (this.object < other.object) return -1
    if (this.object > other.object) return 1
    return 0
  }
}

if (import.meta.vitest) {
  const { describe, it, expect } = import.meta.vitest

  describe('Location', () => {
    it('roundtrip serialization and deserialization (FrozenByteBuffer)', () => {
      const original = new Location(BigInt(1), BigInt(100))
      const serialized = original.serialize()
      const parsed = Location.deserialize(serialized)
      expect(parsed.equals(original)).toBe(true)
    })

    it('roundtrip serialization and deserialization (ByteBuffer)', () => {
      const original = new Location(BigInt(42), BigInt(999))
      const serialized = original.serialize()
      const parsed = Location.deserialize(serialized)
      expect(parsed.equals(original)).toBe(true)
    })

    it('ordering comparison', () => {
      const loc1 = new Location(BigInt(1), BigInt(1))
      const loc2 = new Location(BigInt(1), BigInt(2))
      const loc3 = new Location(BigInt(2), BigInt(1))

      expect(loc1.compare(loc2)).toBeLessThanOrEqual(0)
      expect(loc2.compare(loc3)).toBeLessThanOrEqual(0)
      expect(loc1.compare(loc3)).toBeLessThanOrEqual(0)
      expect(loc2.compare(loc1)).toBe(1)
    })

    it('throws NotEnoughBytesError on insufficient data', () => {
      const buf = new ByteBuffer()
      buf.putVI(BigInt(1)) // missing second value
      const frozen = buf.freeze()
      expect(() => Location.deserialize(frozen)).toThrow(NotEnoughBytesError)
    })

    it('equality check works correctly', () => {
      const a = new Location(BigInt(1), BigInt(1))
      const b = new Location(BigInt(1), BigInt(1))
      const c = new Location(BigInt(2), BigInt(1))

      expect(a.equals(b)).toBe(true)
      expect(a.equals(c)).toBe(false)
    })

    it('handles large 64-bit values', () => {
      const maxVarint = BigInt('4611686018427387903')
      const loc = new Location(maxVarint, maxVarint)
      const serialized = loc.serialize()
      const parsed = Location.deserialize(serialized)
      expect(parsed.equals(loc)).toBe(true)
    })
  })
}
